Egress blog part 3 (#4637)
istio/istio.io: content/blog/2019/egress-traffic-control-in-istio-part-3/index.md

Since operators must configure the applications, this solution is not transparent. Moreover, operators can't use
pod labels or pod service accounts to configure the proxies because the egress proxies don't know about them. Therefore, **the egress proxies are not Kubernetes-aware** and can't fulfill the fourth requirement because
egress proxies cannot enforce policies by source if a Kubernetes artifact specifies the source.
In summary, egress proxies can fulfill the first, second, third and fifth requirements, but can't satisfy the fourth and
the six requirements because they are not transparent and not Kubernetes-aware.

## Advantages of Istio egress traffic control

Istio egress traffic control is **DNS-aware**: you can define policies based on URLs or on wildcard domains like
`*.ibm.com`. In this sense, it is better than Kubernetes network policies which are not DNS-aware.

Istio egress traffic control is **transparent** with regard to TLS traffic, since Istio is transparent:
you do not need to change the applications or to configure their containers.
For HTTP traffic with TLS origination, you must configure the applications to use HTTP instead of HTTPS

> when the applications run with Istio sidecars injected.

Author: I mean "the applications that run as part of Istio service mesh". Maybe it is clear from the context, the stress here is that once you run an application as part of a service mesh (Istio sidecar is injected, manually or automatically), you have to configure that application explicitly to use HTTP instead of HTTPS.

> when the applications run with Istio sidecars injected.

Author: addressed in https://github.com/istio/istio.io/pull/4637/commits/8f7d21ef7e3300a9f4c9a4cc8c885dd6755dbbd9.

Istio egress traffic control is **Kubernetes-aware**: the identity of the source of egress traffic is based on
Kubernetes service accounts. Istio egress traffic control is better than the legacy DNS-aware proxies or firewalls which
are not transparent and not Kubernetes-aware.

Istio egress traffic control is **secure**: it is based on the strong identity of Istio and, when you
apply
[additional security measures](/docs/tasks/traffic-management/egress/egress-gateway/#additional-security-considerations),

> it is tamper-proof.

Author: right, resilient to tampering is less pretentios.

On top of these beneficial features, Istio egress traffic control provides additional advantages:

*  It allows defining access policies in the same language for ingress, egress and in-cluster traffic. You
   need to learn a single policy and configuration language for all types of traffic.
*  Istio egress traffic control is integrated with Istio policy and telemetry adapters and can work out-of-the-box.
*  When you use external monitoring/access control systems with Istio, you must write the adapters for them only once,

>    and then apply the adapters for all types of traffic: ingress, egress and in-cluster.

Author: @rcaballeromx I have doubts regarding "the adapters are applied **to** traffic". The adapters are applied with regard to traffic, or for traffic. For example, apply an adapter to control egress traffic or to perform monitoring of egress traffic.

>    and then apply the adapters for all types of traffic: ingress, egress and in-cluster.

Author: @rcaballeromx It is similar to your suggestion: 
> Use Istio's [traffic management features](/docs/concepts/traffic-management/) for egress traffic:

So you use features **for** egress traffic, apply adapters **for** egress traffic.

>    and then apply the adapters for all types of traffic: ingress, egress and in-cluster.

Author: addressed by https://github.com/istio/istio.io/pull/4637/commits/66d5805078d550e156788c160cd81ecbbb433685 and https://github.com/istio/istio.io/pull/4637/commits/c9366c846077de11b1c3405bc9a6e90a9569161c

*  You can apply Istio traffic management features to egress traffic, such as
   load balancing, passive and active health checking, circuit breaker, timeouts, retries, fault injection, and others.

We call a system that has the advantages above **Istio-aware**.

Let me summarize the features of Istio egress traffic control and of the alternative solutions in the following table:

> | | Kubernetes Network Policies | Legacy Egress Proxy or Firewall | Istio Egress Traffic Control |

Author: Done in https://github.com/istio/istio.io/pull/4637/commits/02e8e322a88213a7719a14e7ccec97e2471bdb02

| --- | --- | --- | ---|

> | DNS-aware | {{< cancel_icon >}} | {{< checkmark_icon >}} | {{< checkmark_icon >}} |

Author: Done in https://github.com/istio/istio.io/pull/4637/commits/02e8e322a88213a7719a14e7ccec97e2471bdb02

alternative solutions.

Let me provide our take on the performance overhead of Istio egress traffic control.
The latency of access to external services could be already high, so adding the overhead
of two or three proxies inside the cluster could be not very significant.
After all, in the microservice architecture you can have chains of dozens of calls between microservices, so adding an
additional hop with one or two proxies in the egress gateway should not have a large impact.

Moreover, we are working to reduce performance overhead of Istio.
Possible optimizations are to extend Envoy to handle wildcard domains so there will be no need for the
third proxy; or to use mutual TLS for authentication only, without encrypting the TLS traffic (since it is already
encrypted).

## Summary

I hope that after reading this series you are convinced that controlling egress traffic is very important for the
security of your cluster.
I also hope that I managed to convince you that Istio can serve as an effective tool for controlling egress traffic
securely, and that Istio has multiple advantages over the alternative solutions.

> In my opinion, you can even choose secure control of egress traffic as the first use case for applying Istio to your

Reviewer: ```suggestion
In my opinion, secure control of egress traffic is a great choice if you are looking for your first Istio use case.
```

> cluster.

Reviewer: ```suggestion

```

> Istio will already be beneficial for you, even before you start using all other features, such as

Reviewer: ```suggestion
In this case, Istio already provides you some benefits even before you start using all other Istio features:
```

Author: @rcaballeromx some benefits does not sound right to me. Istio provides a single benefit of secure control of egress traffic, if it is the first use case.

Author: @rcaballeromx I have applied "some benefits", though I have doubts about it, see https://github.com/istio/istio.io/pull/4637#discussion_r303580907

> traffic management, security, policies and telemetry, applied to traffic between microservices inside the cluster.

Reviewer: Please add links to the concepts for the features you mention here.

Author: Done in https://github.com/istio/istio.io/pull/4637/commits/0dbe862cb8867e1e23cc3e231a4a76be9c3df0b4

> You should pay attention, however, to performance overhead of Istio egress traffic control and measure it for your use

Reviewer: I would remove this sentence. It ends the post on a negative note. You already mentioned the performance considerations and I think it would be better to end on the positive note of all the features Istio provides to readers and how secure control of egress traffic can serve as a starting point.

Author: Done in https://github.com/istio/istio.io/pull/4637/commits/3c3e8aa3fb3dc258470c6eda7aadcb99eebe22ad

In [the first part in the series](/blog/2019/egress-traffic-control-in-istio-part-1/), I presented the attacks involving
egress traffic and the requirements we collected for a secure control system for egress traffic.
In [the second part in the series](/blog/2019/egress-traffic-control-in-istio-part-2/), I presented the Istio way of
securing egress traffic and showed how you can prevent the attacks using Istio.

In this installment, I compare secure control of egress traffic in Istio with alternative solutions such as Kubernetes
Network Policies and legacy egress proxies/firewalls. Finally, I describe the performance considerations regarding the
secure control of egress traffic in Istio.

## Alternative solutions for egress traffic control

First, let me remind you the [requirements for egress traffic control we collected](/blog/2019/egress-traffic-control-in-istio-part-1/#requirements-for-egress-traffic-control):

1.  Support for [TLS](https://en.wikipedia.org/wiki/Transport_Layer_Security) with
    [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication) or for [TLS origination](/docs/reference/glossary/#tls-origination) by Istio.

> 1.  **Monitor** SNI and the source workload of every egress access.

Author: @rcaballeromx Please note that I use the imperative in other bullets: support, monitor, prevent, enforce

1.  Define and enforce **policies per cluster**.
1.  Define and enforce **policies per source**, _Kubernetes-aware_.
1.  **Prevent tampering**.
1.  Traffic control is **transparent** to the applications.

> Now, let me present two alternative solutions for egress traffic control and show which requirements they satisfy, and,

Author: @rcaballeromx I added "namely Kubernetes Network Policies and egress proxies and firewalls" in https://github.com/istio/istio.io/pull/4637/commits/3e7ef7662f8ceda57a1bec3a1340081bac2f7f0a

more importantly, cannot satisfy.

> The most natural solution for egress traffic control is

Author: @rcaballeromx Kubernetes Network Policies is the "native" Kubernetes solution. For a person familiar with Kubernetes, this would be the first mechanism to come to mind in the context of traffic control. However, it does not fulfill all the requirements. I am not trying to "sell Istio", and as part of my blog post I present the alternative solutions. This section is an answer to the questions I encountered when presenting egress traffic control in Istio: how is it better than Kubernetes Network Policies? why should I use Istio for egress traffic control and not a DNS-aware egress proxy?

> The most natural solution for egress traffic control is

Author: Addressed by https://github.com/istio/istio.io/pull/4637/commits/1dbe203d93535b322fca096db0a2a5a20991ca3b.

[Kubernetes Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/). Using
Kubernetes Network Policies, cluster operators can specify which external services can be accessed by which pods. The
pods may be identified by pod labels, namespace labels, or by IP ranges. The external services can be specified by IP

> ranges: Kubernetes Network Policies are not DNS-aware. The first requirement is satisfied since any TCP traffic can be

Author: addressed by cd31534771739781c5b3bcae63f37454fd5d4ad9 and 3472526c5656c4ef4eeaf99b3e4d1ed4a421fd24

> controlled by Kubernetes Network Policies. The requirements 3 and 4 are satisfied partially: the policies can be

Author: done in cf85e7b

> specified per cluster or per pod, however the external services cannot be identified by domain names.

Author: done in cf85e7b910519df608bd2cfcc5e1ba99091ec1b1

The fifth requirement is satisfied if the attackers are not able to break from a malicious container into the Kubernetes
node and to interfere with the implementation of the Kubernetes Network Policies inside the node.
The sixth requirement is satisfied as well: there is no need to change the code or the
container environment. We can say that Kubernetes Network Policies provide transparent, Kubernetes-aware egress traffic
control, which is not DNS-aware.

Another approach that predates Kubernetes network policies is a **DNS-aware egress proxy/firewall**. In this
approach, applications are configured to direct the traffic to the proxy and to use some proxy protocol, e.g.
[SOCKS](https://en.wikipedia.org/wiki/SOCKS).
Since the applications must be specially configured, this solution is not transparent. Moreover, egress proxies are not

> Kubernetes-aware, since neither pod labels nor pod service account are known to the egress proxy. Such egress proxies

Author: @rcaballeromx Good idea to use operators as the subjects. Good that you did not use **cluster** operators, since the egress proxies are not a Kubernetes concept, so Kubernetes cluster operators not necessarily configure them, some other operators do.

Accessing external services already could have high latency and the overhead added
because of two or three proxies inside the cluster could likely not be very significant by comparison.
After all, applications with a microservice architecture can have chains of dozens of calls between microservices.
Therefore, an additional hop with one or two proxies in the egress gateway should not have a large impact.

Moreover, we continue to work towards reducing Istio's performance overhead.
Possible optimizations include:

- Extending Envoy to handle wildcard domains: This would eliminate the need for a third proxy between
  the application and the external services for that use case.
- Using mutual TLS for authentication only without encrypting the TLS traffic, since the traffic is already
  encrypted.

## Summary

I hope that after reading this series you are convinced that controlling egress traffic is very important for the
security of your cluster.
Hopefully, I also managed to convince you that Istio is an effective tool to control egress traffic
securely, and that Istio has multiple advantages over the alternative solutions.

> If you need to control egress traffic in a secure and transparent way, specifying external services as domain names and

Reviewer: ```suggestion
Istio is the only solution I'm aware of that lets you:

- Control egress traffic in a secure and transparent way
- Specify external services as domain names

```

> specifying traffic source by Kubernetes artifacts, Istio is the only solution for you I am aware of.

Reviewer: ```suggestion
- Use Kubernetes artifacts to specify a traffic source.
```

In my opinion, secure control of egress traffic is a great choice if you are looking for your first Istio use case.
In this case, Istio already provides you some benefits even before you start using all other Istio features:
[traffic management](/docs/tasks/traffic-management/), [security](/docs/tasks/security/),
[policies](/docs/tasks/policy-enforcement/) and [telemetry](/docs/tasks/telemetry/), applied to traffic between
microservices inside the cluster.

> So, if you did not yet have a chance to work with Istio, [install](/docs/setup/kubernetes/install/) it in your cluster

Reviewer: ```suggestion
So, if you haven't had the chance to work with Istio yet, [install Istio](/docs/setup/kubernetes/install/) on your cluster
```

> and check [egress traffic control](/docs/tasks/traffic-management/egress/) and other

Reviewer: ```suggestion
and check our [egress traffic control task](/docs/tasks/traffic-management/egress/) and the tasks for the other
```

> [Istio features](https://istio.io/docs/tasks/). Tell us what you think at [discuss.istio.io](https://discuss.istio.io).

Reviewer: ```suggestion
[Istio features](https://istio.io/docs/tasks/). We also want to hear from you, please join us at [discuss.istio.io](https://discuss.istio.io).
```

The following table summarizes the egress traffic control features that Istio and the alternative solutions provide:

| | Istio Egress Traffic Control | Kubernetes Network Policies | Legacy Egress Proxy or Firewall |
| --- | --- | --- | ---|
| DNS-aware | {{< checkmark_icon >}} | {{< cancel_icon >}} | {{< checkmark_icon >}} |
| Kubernetes-aware | {{< checkmark_icon >}} | {{< checkmark_icon >}} | {{< cancel_icon >}} | {
| Transparent | {{< checkmark_icon >}} | {{< checkmark_icon >}} | {{< cancel_icon >}} |
| Istio-aware | {{< checkmark_icon >}} | {{< cancel_icon >}} | {{< cancel_icon >}} |

## Performance considerations

Controlling egress traffic using Istio has a price: increased latency of calls to external services and
increased CPU usage by the cluster's pods.
Traffic passes through two proxies:

- The application's sidecar proxy
- The egress gateway's proxy

If you use [TLS egress traffic to wildcard domains](/docs/tasks/traffic-management/egress/wildcard-egress-hosts/),
you must add
[an additional proxy](/docs/tasks/traffic-management/egress/wildcard-egress-hosts/#wildcard-configuration-for-arbitrary-domains),

> adding a third proxy between the application and the external service. Since the traffic between the egress gateway's proxy and

Author: Done in https://github.com/istio/istio.io/pull/4637/commits/e7924cced9b20141bbf6a3736028099e4685b4e8

Therefore, an additional hop with one or two proxies in the egress gateway should not have a large impact.

Moreover, we continue to work towards reducing Istio's performance overhead.
Possible optimizations include:

- Extending Envoy to handle wildcard domains: This would eliminate the need for a third proxy between
  the application and the external services for that use case.
- Using mutual TLS for authentication only without encrypting the TLS traffic, since the traffic is already
  encrypted.

## Summary

I hope that after reading this series you are convinced that controlling egress traffic is very important for the
security of your cluster.
Hopefully, I also managed to convince you that Istio is an effective tool to control egress traffic
securely, and that Istio has multiple advantages over the alternative solutions.
In my opinion, secure control of egress traffic is a great choice if you are looking for your first Istio use case.
In this case, Istio already provides you some benefits even before you start using all other Istio features:
[traffic management](/docs/tasks/traffic-management/), [security](/docs/tasks/security/),

> [policies](/docs/tasks/policy-enforcement/) and [telemetry](/docs/tasks/telemetry/), applied to traffic between microservices inside the cluster.

Reviewer: @vadimeisenbergibm What is the final message you want to give readers after they have followed you through all three parts? The message on this finishing paragraph is a bit muddled. Try to picture someone who followed the entire journey. They learned what controlling egress traffic securely was all about, the different attacks that make it important, and the different ways they can protect against said attacks. After all of that, you reached the conclusion that Istio's secure control of egress traffic is the way to go and explained why. 
IMO, this paragraph should communicate three things to your readers:
1. What's next? Assume readers have followed along and secured their egress traffic, what specific actions would you recommend they take next?
2. Where can I learn more? I think this is what you were trying to do by listing the features. However, those still feel disconnected from the topic at hand. What tasks do you recommend they complete next? What reference materials will allow your readers to do more around traffic management?
3. Summarize the value that Istio gives them and that they cannot get any other way.
Some of this things are alluded to so you might only need to rearrange and reword a little.

Author: Addressed in https://github.com/istio/istio.io/pull/4637/commits/f2a4c183b96d82deed21f3576e46b220063c9deb and https://github.com/istio/istio.io/pull/4637/commits/667c470c82b0a25c882b12c5b40d02b11647a7b6
